porting/cpphdrcheck.t: test perl's headers with C++ compilers

tonycoz opened the pull request:
This searches for a C++ compiler based on the supplied C compiler, and
checks that compiler for any options controlling the C++ standard
requested, including simple checks that the compiler supports that
standard.

If a C++ compiler is found, test compilation of the same simple code
as above but with the perl headers included after any C++ headers.

Ideally we'd also test runtime, but would require more complex test
code, which I leave to later contributors (which may still be me).

Tested at various times with:

 - MSVC
 - gcc
 - clang
 - Oracle/Sun Development Workshop cc (CC is the C++ compiler), on Oracle Linux
 - Intel oneAPI compiler (llvm based apparently, and now free to use)
 - Intel classic compiler (discontinued)

Currently this probes the compiler for C++ sanity with the perl
ccflags, since icc (Intel classic) would successfully build the sample
without perl's ccflags, but then fail with both the headers and perl's
ccflags.  It turned out to fail with just the ccflags, and since the
primary intent is to test the headers, I probe *with* ccflags.

The Sun Workshop compiler failed to build the C++11 or 14 sample at
all in my testing, which may have been due to an installation problem.

This branch also removes an option added to ccflags by the cygwin hints
that prevented C++ code from compiling.

tonycoz committed:
porting/cpphdrcheck.t: test perl's headers with C++ compilers

This searches for a C++ compiler based on the supplied C compiler, and
checks that compiler for any options controlling the C++ standard
requested, including simple checks that the compiler supports that
standard.

If a C++ compiler is found, test compilation of the same simple code
as above but with the perl headers included after any C++ headers.

Ideally we'd also test runtime, but would require more complex test
code, which I leave to later contributors (which may still be me).

Tested at various times with:

 - MSVC
 - gcc
 - clang
 - Oracle/Sun Development Workshop cc (CC is the C++ compiler), on
   Oracle Linux
 - Intel oneAPI compiler (llvm based apparently, and now free to use)
 - Intel classic compiler (discontinued)

Currently this probes the compiler for C++ sanity with the perl
ccflags, since icc (Intel classic) would successfully build the sample
without perl's ccflags, but then fail with both the headers and perl's
ccflags.  It turned out to fail with just the ccflags, and since the
primary intent is to test the headers, I probe *with* ccflags.

The Sun Workshop compiler failed to build the C++11 or 14 sample at
all in my testing, which may have been due to an installation problem.

tonycoz committed:
hints/cygwin.sh: don't touch system symbol __STRICT_ANSI__

Adding _GNU_SOURCE, which was done much later, should give us the
GNU-ish symbols (though POSIX_C_SOURCE would be more general.)

I couldn't find the reason for this being added, but the C++
headers react badly to it:

In file included from /usr/lib/gcc/x86_64-pc-cygwin/11/include/c++/iostream:38,
                 from source.cpp:1:
/usr/lib/gcc/x86_64-pc-cygwin/11/include/c++/x86_64-pc-cygwin/bits/c++config.h:573:2: warning: #warning "__STRICT_ANSI__ seems to have been undefined; this is not supported" [-Wcpp]
  573 | #warning "__STRICT_ANSI__ seems to have been undefined; this is not supported"
      |  ^~~~~~~
In file included from /usr/lib/gcc/x86_64-pc-cygwin/11/include/c++/bits/max_size_type.h:37,
                 from /usr/lib/gcc/x86_64-pc-cygwin/11/include/c++/bits/ranges_base.h:38,
                 from /usr/lib/gcc/x86_64-pc-cygwin/11/include/c++/string_view:48,
                 from /usr/lib/gcc/x86_64-pc-cygwin/11/include/c++/bits/basic_string.h:48,
                 from /usr/lib/gcc/x86_64-pc-cygwin/11/include/c++/string:55,
                 from /usr/lib/gcc/x86_64-pc-cygwin/11/include/c++/bits/locale_classes.h:40,
                 from /usr/lib/gcc/x86_64-pc-cygwin/11/include/c++/bits/ios_base.h:41,
                 from /usr/lib/gcc/x86_64-pc-cygwin/11/include/c++/ios:42,
                 from /usr/lib/gcc/x86_64-pc-cygwin/11/include/c++/ostream:38,
                 from /usr/lib/gcc/x86_64-pc-cygwin/11/include/c++/iostream:39,
                 from source.cpp:1:
/usr/lib/gcc/x86_64-pc-cygwin/11/include/c++/numbers:139:9: error: unable to find numeric literal operator 'operator""Q'
  139 |       = 2.718281828459045235360287471352662498Q;
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
...

jkeenan commented:
1. I fetched this pull request today, rebased it on (post-perl-5.40.0) blead and compiled and tested it on FreeBSD-13 with both gcc+12 and clang++14.  All tests PASSed.
2. Several weeks ago I forked a 'smoke-me' branch from this pull request.  It did very well on two smoke-testing rigs (http://perl.develop-help.com/?b=smoke-me%2Ftonyc%2Fcpp-header-porting) -- but I think the p.r. should still get some eyeballs from people who know more about C++ and headers than I do.

jkeenan commented:
> 1. I fetched this pull request today, rebased it on (post-perl-5.40.0) blead and compiled and tested it on FreeBSD-13 with both gcc+12 and clang++14.  All tests PASSed.
> 
>     2. Several weeks ago I forked a 'smoke-me' branch from this pull request.  It did very well on two smoke-testing rigs (http://perl.develop-help.com/?b=smoke-me%2Ftonyc%2Fcpp-header-porting) -- but I think the p.r. should still get some eyeballs from people who know more about C++ and headers than I do.

Is there anyone familiar enough with C++ headers to give this proposal for a porting test a review?

IMO it's the sort of thing that, if we're going to pull it into core, we should do so now, at an early point in the current development cycle.

mauke commented on t/porting/cpphdrcheck.t:
```diff
@@ -0,0 +1,580 @@
+# we chdir around a bit below, which breaks relative paths and Carp
+@INC = map File::Spec->rel2abs($_), @INC;
+
+# the intent is the compiler detection done here will move into a module
```
Missing punctuation between this line and the next?

mauke commented on t/porting/cpphdrcheck.t:
```diff
@@ -0,0 +1,580 @@
+
+# the intent is the compiler detection done here will move into a module
+# EU::CB doesn't provide what I need here, EU::CppGuess does have some of
+# it but isn't core, and has it's own limitations
```
`s/it's/its/`

mauke commented on t/porting/cpphdrcheck.t:
```diff
@@ -0,0 +1,580 @@
+note @$diag;
+
+$ccpp_cfg
+  or skip_all("Cannot find a C++ compiler corresponding to $Config{cc}");
```
I think that should be `$cc`, not `$Config{cc}`.

mauke commented on t/porting/cpphdrcheck.t:
```diff
@@ -0,0 +1,580 @@
+$ccpp_cfg
+  or skip_all("Cannot find a C++ compiler corresponding to $Config{cc}");
+
+note join "\n", @$diag;
```
Why are we noting `@$diag` again? (cf. line 36)

mauke commented on t/porting/cpphdrcheck.t:
```diff
@@ -0,0 +1,580 @@
+    }
+    # msvc
+    # may need work if we ever support clang-cl
+    elsif ($ccarg =~ m!([\\/]|^)\bcl(?:(?:\Q$exe\E)?)?$!i
```
This `\b` is technically redundant here.
Why the doulbe `(?: )` around `$exe`?

tonycoz replied:
IIRC I was thinking of handling clang-cl here and didn't clean up after myself.

mauke commented on t/porting/cpphdrcheck.t:
```diff
@@ -0,0 +1,580 @@
+        }
+
+        if ($ver =~ /Intel(?:\(R\))? (?:.*)C.* Compiler/) {
+            if (($ccpp = $ccarg) =~ s/\bicc((?:.exe)?)$/icpc$1/iaa
```
Why isn't the `.` in `.exe` escaped?

tonycoz commented:
Fixed everything I think:
- added a comma
- removed a `'`
- switched to `$cc`
- removed duplicate diagnostic output on the initial detection
- used `\Q$exe\E` where I was using `.exe`
- cleaned up the MSVC regexp
